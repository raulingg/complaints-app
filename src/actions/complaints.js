import database from '../firebase/firebase';

export const addComplaint = complaint => ({
  type: 'ADD_COMPLAINT',
  complaint,
});

export const startAddComplaint = (complaintData = {}) => (dispatch, getState) => {
  const {
    auth: { uid },
  } = getState();
  const {
    title = '',
    content = '',
    reportedAt = 0,
    happenedAt = 0,
    reportTo = '',
    address = '',
  } = complaintData;
  const complaint = { title, content, reportedAt, happenedAt, reportTo, address, uid };

  return database
    .ref('complaints')
    .push(complaint)
    .then(ref => {
      dispatch(
        addComplaint({
          id: ref.key,
          ...complaint,
        })
      );
    });
};

export const editComplaint = (id, updates) => ({
  type: 'EDIT_COMPLAINT',
  id,
  updates,
});

export const startEditComplaint = (id, updates) => dispatch =>
  database
    .ref(`complaints/${id}`)
    .update(updates)
    .then(() => {
      dispatch(editComplaint(id, updates));
    });

export const setComplaints = complaints => ({
  type: 'SET_COMPLAINTS',
  complaints,
});

export const startSetComplaints = () => dispatch =>
  database
    .ref('complaints')
    .once('value')
    .then(snapshot => {
      const complaints = [];

      snapshot.forEach(childSnapshot => {
        complaints.push({
          id: childSnapshot.key,
          ...childSnapshot.val(),
        });
      });

      dispatch(setComplaints(complaints));
    });
