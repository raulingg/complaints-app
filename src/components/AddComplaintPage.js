import React from 'react';
import ComplaintFormComponent from './ComplaintForm';

const AddComplaintPage = () => (
  <div>
    <h1>AÃ±adir nuevo complaint </h1>
    <ComplaintFormComponent />
  </div>
);

export default AddComplaintPage;
