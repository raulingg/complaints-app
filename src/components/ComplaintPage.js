import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';

export const ComplaintPage = ({ complaint }) => (
  <div>
    <h1>{complaint.title}</h1>
    <p>
      Denuncia: <span>{complaint.reportTo}</span>
    </p>
    <div dangerouslySetInnerHTML={{ __html: complaint.content }} />
    <p>
      <span>Denunciado el {moment(complaint.happenedAt).format()}</span>
    </p>
  </div>
);

const mapStateToProps = (state, props) => ({
  complaint: state.complaints.find(complaint => String(complaint.id) === props.match.params.id),
});

export default connect(mapStateToProps)(ComplaintPage);
