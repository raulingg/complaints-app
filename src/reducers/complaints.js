// Expenses Reducer

const complaintsReducerDefaultState = [];

export default (state = complaintsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_COMPLAINT':
      return [...state, action.complaint];
    case 'REMOVE_COMPLAINT':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_COMPLAINT':
      return state.map(complaint => {
        if (complaint.id === action.id) {
          return {
            ...complaint,
            ...action.updates,
          };
        }
        return complaint;
      });
    case 'SET_COMPLAINTS':
      return action.complaints;
    default:
      return state;
  }
};
