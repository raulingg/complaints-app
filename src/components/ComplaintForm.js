import React, { Component } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { startAddComplaint as startAddComplaintAction } from '../actions/complaints';

export class ComplaintForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: props.complaint ? props.complaint.title : '',
      content: props.complaint ? props.complaint.content : '',
      reportTo: props.complaint ? props.complaint.reportTo : '',
      happenedAt: props.complaint ? moment(props.complaint.happenedAt) : moment(),
    };
  }

  onTitleChange = e => {
    const title = e.target.value;
    this.setState(() => ({ title }));
  };

  onContentChange = e => {
    const content = e.target.value;
    this.setState(() => ({ content }));
  };

  onReportToChange = e => {
    const reportTo = e.target.value;
    this.setState(() => ({ reportTo }));
  };

  onHappenedAtChange = e => {
    const happenedAt = e.target.value;
    this.setState(() => ({ happenedAt }));
  };

  onSubmit = e => {
    e.preventDefault();
    const { title, content, happenedAt, reportTo } = this.state;
    const { startAddComplaint, history } = this.props;
    startAddComplaint({
      title,
      content,
      happenedAt,
      reportTo,
    });
    history.push('/');
  };

  render() {
    const { title, content, happenedAt, reportTo } = this.state;
    return (
      <form onSubmit={this.onSubmit}>
        <div>
          <input type="text" name="title" value={title} onChange={this.onTitleChange} />
        </div>
        <div>
          <textarea
            name="content"
            maxLength="2000"
            rows="10"
            value={content}
            onChange={this.onContentChange}
          />
        </div>
        <div>
          <input
            type="date"
            name="happenedAt"
            value={happenedAt}
            onChange={this.onHappenedAtChange}
          />
        </div>
        <div>
          <input type="text" name="reportTo" value={reportTo} onChange={this.onReportToChange} />
        </div>
        <div>
          <button className="button">Guardar cambios</button>
        </div>
      </form>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddComplaint: complaint => dispatch(startAddComplaintAction(complaint)),
});

export default connect(
  undefined,
  mapDispatchToProps
)(ComplaintForm);
