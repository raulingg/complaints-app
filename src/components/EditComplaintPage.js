import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { compose } from 'recompose';
import ComplaintFormComponent from './ComplaintForm';
import { startEditComplaint as startEditComplaintAction } from '../actions/complaints';
import withAuthorization from './auth/withAuthorization';

export class EditComplaintPage extends Component {
  onSubmit = updates => {
    const {
      history,
      startEditComplaint,
      complaint: { id },
    } = this.props;
    startEditComplaint(id, updates);
    history.push('/');
  };

  render() {
    const { complaint } = this.props;
    return (
      <div>
        <h1>Editar denuncia </h1>
        <ComplaintFormComponent complaint={complaint} onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  complaint: state.complaints.find(complaint => String(complaint.id) === props.match.params.id),
});

const mapDispatchToProps = dispatch => ({
  startEditComplaint: (id, complaint) => dispatch(startEditComplaintAction(id, complaint)),
});

export default compose(
  withRouter,
  withAuthorization,
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)(EditComplaintPage);
