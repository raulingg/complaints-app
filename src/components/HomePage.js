import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Button from '../styles/components/Buttons';
import { startSetComplaints as startSetComplaintsAction } from '../actions/complaints';
import LoadingPage from './LoadingPage';

export class HomePage extends Component {
  state = {
    isLoading: true,
  };

  async componentDidMount() {
    const { startSetComplaints } = this.props;
    await startSetComplaints();
    this.setState({ isLoading: false });
  }

  render() {
    const { complaints } = this.props;
    const { isLoading } = this.state;

    return isLoading ? (
      <LoadingPage />
    ) : (
      <div>
        <div>
          <div>Denuncia</div>
          <div>Tipo</div>
          <div>Fecha</div>
        </div>
        <div>
          <Button as={Link} to="/add">
            Denunciar
          </Button>
          {complaints.length === 0 ? (
            <div>
              <span>No encontramos ninguna denuncia.</span>
            </div>
          ) : (
            complaints.map(complaint => (
              <div key={complaint.id}>
                <p>
                  <Link to={`show/${complaint.id}`}>{complaint.title}</Link>
                </p>
              </div>
            ))
          )}
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  complaints: state.complaints,
});

const mapDispatchToProps = dispatch => ({
  startSetComplaints: () => dispatch(startSetComplaintsAction()),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(HomePage);
