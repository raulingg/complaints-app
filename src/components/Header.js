import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { startLogout as startLogoutAction } from '../actions/auth';
import * as HeaderStyles from '../styles/components/Header';
import ContentContainer from '../styles/components/ContentContainer';
import { ButtonLink } from '../styles/components/Buttons';

export const Header = ({ isAuthenticated, startLogout }) => (
  <HeaderStyles.Header>
    <ContentContainer>
      <HeaderStyles.Content>
        <HeaderStyles.Title to="/">
          <h1>Denuncias</h1>
        </HeaderStyles.Title>
        {isAuthenticated ? (
          <ButtonLink onClick={startLogout}>Salir</ButtonLink>
        ) : (
          <ButtonLink as={Link} to="/login">
            Iniciar sesi√≥n
          </ButtonLink>
        )}
      </HeaderStyles.Content>
    </ContentContainer>
  </HeaderStyles.Header>
);

const mapStateToProps = state => ({
  isAuthenticated: !!state.auth.user,
});

const mapDispatchToProps = dispatch => ({
  startLogout: () => dispatch(startLogoutAction()),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Header);
