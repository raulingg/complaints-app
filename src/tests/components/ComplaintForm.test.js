import React from 'react';
import { shallow } from 'enzyme';
import { ComplaintForm } from '../../components/ComplaintForm';
import complaints from '../fixtures/complaints';
import moment from '../__mocks__/moment';

describe('RENDER', () => {
  test('should render ComplaintForm correctly', () => {
    const wrapper = shallow(<ComplaintForm />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should render ComplaintForm with complaint data', () => {
    const wrapper = shallow(<ComplaintForm complaint={complaints[0]} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('EVENTS', () => {
  test('should set title state when the title input changes', () => {
    const value = 'New title';
    const wrapper = shallow(<ComplaintForm />);
    wrapper.find({ name: 'title' }).simulate('change', {
      target: { value },
    });
    expect(wrapper.state('title')).toBe(value);
  });

  test('should set content state when the content input changes', () => {
    const value = 'New content';
    const wrapper = shallow(<ComplaintForm />);
    wrapper.find({ name: 'content' }).simulate('change', {
      target: { value },
    });
    expect(wrapper.state('content')).toBe(value);
  });

  test('should set reportTo state when the reportTo input changes', () => {
    const value = 'Angelica merina';
    const wrapper = shallow(<ComplaintForm />);
    wrapper.find({ name: 'reportTo' }).simulate('change', {
      target: { value },
    });
    expect(wrapper.state('reportTo')).toBe(value);
  });

  test('should set happenedAt state when the happenedAt input changes', () => {
    const value = '2018-12-31';
    const wrapper = shallow(<ComplaintForm />);
    wrapper.find({ name: 'happenedAt' }).simulate('change', {
      target: { value },
    });
    expect(wrapper.state('happenedAt')).toBe(value);
  });

  test('should call startAddComplaint action and history function when form is submitted', () => {
    const startAddComplaintSpy = jest.fn();
    const historySpy = {
      push: jest.fn(),
    };

    const wrapper = shallow(
      <ComplaintForm
        complaint={complaints[0]}
        startAddComplaint={startAddComplaintSpy}
        history={historySpy}
      />
    );
    wrapper.find('form').simulate('submit', {
      preventDefault: () => {},
    });
    expect(startAddComplaintSpy).toHaveBeenCalledWith({
      title: complaints[0].title,
      content: complaints[0].content,
      reportTo: complaints[0].reportTo,
      happenedAt: moment(complaints[0].happenedAt),
    });
    expect(historySpy.push).toHaveBeenCalledTimes(1);
  });
});
