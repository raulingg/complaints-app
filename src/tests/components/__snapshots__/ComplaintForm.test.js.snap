// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RENDER should render ComplaintForm correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <input
      name="title"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div>
    <textarea
      maxLength="2000"
      name="content"
      onChange={[Function]}
      rows="10"
      value=""
    />
  </div>
  <div>
    <input
      name="happenedAt"
      onChange={[Function]}
      type="date"
      value={"1970-01-01T00:00:00.000Z"}
    />
  </div>
  <div>
    <input
      name="reportTo"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div>
    <button
      className="button"
    >
      Guardar cambios
    </button>
  </div>
</form>
`;

exports[`RENDER should render ComplaintForm with complaint data 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <input
      name="title"
      onChange={[Function]}
      type="text"
      value="Me robaron mi celular"
    />
  </div>
  <div>
    <textarea
      maxLength="2000"
      name="content"
      onChange={[Function]}
      rows="10"
      value="esta yo en la calle cuando de repente un maricon me robo el celular."
    />
  </div>
  <div>
    <input
      name="happenedAt"
      onChange={[Function]}
      type="date"
      value={"1970-01-01T00:00:01.023Z"}
    />
  </div>
  <div>
    <input
      name="reportTo"
      onChange={[Function]}
      type="text"
      value="Americatel SAC"
    />
  </div>
  <div>
    <button
      className="button"
    >
      Guardar cambios
    </button>
  </div>
</form>
`;
